internal/oci/runtime_oci.go

```diff
@@ -832,50 +869,68 @@ func (r *runtimeOCI) StopLoopForContainer(c *Container) {
 	done := make(chan struct{})
 	go func() {
 		for {
-			if err := c.Living(); err != nil {
+			if _, err := c.Living(); err != nil {
 				// The initial container process either doesn't exist, or isn't ours.
 				if !errors.Is(err, ErrNotFound) {
 					log.Warnf(ctx, "Failed to find process for container %s: %v", c.ID(), err)
 				}
 				close(done)
 				return
 			}
-			// the PID is still active and belongs to the container, continue to wait
-			time.Sleep(100 * time.Millisecond)
+
+			// The PID is still active and belongs to the container, continue to wait.
+			time.Sleep(stopProcessWatchSleep)
 		}
 	}()
 
 	// Operate in terms of targetTime, so that we can pause in the middle of the operation
 	// to catch a new timeout (and possibly ignore that new timeout if it's not correct to
 	// take a new one).
-	targetTime := time.Unix(1<<50-1, 0)
-	for finished := false; !finished; {
+	targetTime := time.Now().AddDate(+1, 0, 0) // A year from this one.
+
+	blockedTimer := time.AfterFunc(stopProcessBlockedInternal, func() {
+		if state, err := c.Living(); err == nil && state == "D" {
```

Comment:
I'm not sure that this happens *only* on uninterruptible sleep (it's likely only on uninterruptible sleep, but in another customer case under investigation, I could not find evidence that the process in question was in uninterruptible sleep). Anyway, this state is for the runtime's init process, right?
Nit: "Detected process (%d) blocked in uninterruptible sleep for more than %d seconds for container: %s" would IMO make it clearer, especially given that you test for "D"

Reply:
@andreaskaris, most of the time, we only track the init PID within the container within CRI-O, so if the process is blocked or waiting either for one of its children or a thread (albeit, this case is slightly different), then you could also see things blocked despite the primary process within the container sleeping with "S" state, etc.

We can add a message that spells out a list of children and threads of the init process with their state, which could help a little more. This is something that I ran through @mrunalp already, but it's also something he wants to include as a later, perhaps follow-up, change.

Reply:
[...]
> With the test for "D" only: "Detected process (%d) blocked **in uninterruptible sleep** for more than %d seconds for container: %s" would IMO make it clearer, especially given that you test for "D"

Done.

Reply:
thanks!